sdk/include/reactos/shellutils.h

```diff
@@ -88,6 +88,42 @@ inline BOOL _ROS_FAILED_HELPER(HRESULT hr, const char* expr, const char* filenam
 } /* extern "C" */
 #endif /* defined(__cplusplus) */
 
+static inline UINT
+SHELL_ErrorBoxHelper(HWND hwndOwner, UINT Error)
+{
+    WCHAR buf[400];
+    UINT cch, msgId, u32_errstr = 2;
+
+    if (!IsWindowVisible(hwndOwner))
+        hwndOwner = NULL;
+    if (Error == ERROR_SUCCESS)
+        Error = ERROR_INTERNAL_ERROR;
+
+    msgId = Error;
+retry:
+    cch = FormatMessageW(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+                         NULL, msgId, 0, buf, _countof(buf), NULL);
+    if (!cch)
+    {
+        if (HIWORD(msgId) == HIWORD(HRESULT_FROM_WIN32(1)))
+        {
+            msgId = HRESULT_CODE(msgId); // Extract ERROR_ from HRESULT_FROM_WIN32
+            goto retry;                  // and try again.
+        }
+        cch = LoadStringW(LoadLibraryW(L"USER32"), u32_errstr, buf, _countof(buf));
+        wsprintfW(buf + cch, L"\n\n%#x (%d)", Error, Error);
+    }
+    MessageBoxW(hwndOwner, buf, NULL, MB_OK | MB_ICONSTOP);
+    return Error;
+}
+#ifdef __cplusplus
+template<class H> static UINT
+SHELL_ErrorBox(H hwndOwner, UINT Error = GetLastError())
```

Comment:
> What is the purpose of this template?

 > And before anyone asks, it is a template to discourage anyone from passing NULL as the HWND.

Reply:
I want to discourage people from writing `SHELL_ErrorBox(NULL, error);`. Calling `SHELL_ErrorBox(hSomeWindow, error);` where `hSomeWindow` turns out to be NULL at run-time is acceptable.

Reply:
If people do really intend to do that, then they will simply write MessageBox(NULL,... instead.
I do also think it is overengineered trying to prevent that. Like hbelusca said: passing NULL there is technically legit code.
It being wrong poor GUI design is true in most cases, but I would discrecommend trying to forbid that like you attempted here.

Reply:
> then they will simply write MessageBox(NULL,... instead

But with which message string? `MessageBox(NULL, L"Error blah blah", ...)` will get you a comment in the PR because this already happened to me =).

You are not supposed to try to display UI without a owner handle inside the shell, that is the point.  Even functions designed to be called by Rundll32.exe has a HWND parameter you can use.

If you are deep inside some COM object you can call `IUnknown_GetWindow`. Otherwise there should not really be any place where you are allowed to `MessageBox` but nobody passed you a `HWND`.

Reply:
I don't think we should make this code more ugly than necessary. It's not worth it just to clap on other peoples fingers.
I do vote for stripping that template.